### PR TITLE
Add grpo trainer

### PR DESCRIPTION
**Pull request by LSX-Sneakerprogrammer**
This is an attempt to add grpo trainer.

It has been shown in qwen2.5-math and deepseek-math that grpo can provide effective help for math tasks. Therefore, this PR aims to implement the grpo trainer based on the content in technical report and ppo trainer in OpenRLHF repository.

The main ideas of this PR are:

* Implement grpo trainer.
* Add grpo_utils, including new experience_maker ```GRPOExperienceMaker``` and new replay buffer ```GRPOReplayBuffer``` supporting for grpo trainer.
* the ```GRPOExperience``` stores all repsonses of one prompt, which is different from ppo trainer. Therefore, the ```GRPOExperienceMaker``` and ```GRPOReplayBuffer``` are needed.

I have done a testing experiment for this PR, which is recorded in: https://wandb.ai/929170655/openrlhf_train_grpo?nw=nwuser929170655.

The testing experiment is based on **Actor Model: Qwen2.5-7B-Instruct**，**Reward Model: Qwen2.5-Math-RM-72B**, **Dataset: 10k prompts from orca-math-word-problems-200k**, with scripts like:
```python
set -x

read -r -d '' training_commands <<EOF
openrlhf.cli.train_grpo \
   --pretrain Qwen/Qwen2.5-7B-Instruct \
   --remote_rm_url xxxx \ # api server url of Qwen/Qwen2.5-Math-RM-72B
   --save_path ./checkpoint/qwen2_5-7b-grpo \
   --save_steps -1 \
   --logging_steps 1 \
   --eval_steps -1 \
   --n_responses 8 \ # how many responses generated for one prompt
   --micro_train_batch_size 8 \
   --train_batch_size 128 \
   --micro_rollout_batch_size 8 \
   --rollout_batch_size 256 \
   --max_epochs 1 \
   --prompt_max_len 1024 \
   --generate_max_len 1024 \
   --zero_stage 2 \
   --bf16 \
   --actor_learning_rate 5e-7 \
   --init_kl_coef 0.01 \
   --prompt_data  orca-math-word-problems-200k \
   --input_key question\
   --apply_chat_template \
   --max_samples 10000 \
   --normalize_reward \
   --adam_offload \
   --flash_attn \
   --load_checkpoint \
   --gradient_checkpointing
EOF

if [[ ${1} != "slurm" ]]; then
    deepspeed --module $training_commands
fi
```

The ray version of grpo is currently under development and will be submitted soon.

Thanks for your time on reviewing this PR :)

### CODE REVIEW

**Comment by hijkzzz** [comment]
It seems that most of the code is copied from train_ppo.py. The difference between GRPO and PPO is just removing the value net and calculating the advantage only use rewards. Would it be possible to reuse the flow of train_ppo.py by adding a `--use_grpo` option?

**Comment by LSX-Sneakerprogrammer** [comment]
> It seems that most of the code is copied from `train_ppo.py`. The difference between GRPO and PPO is just removing the value net and calculating the advantage only use rewards. Would it be possible to reuse the flow of train_ppo.py by adding a `--use_grpo` option?

Hi, thanks for your reviewing. I have merged ```grpo_train.py``` and ```ppo_train.py```,  ```grpo_trainer.py``` and ```ppo_trainer.py```. Besides, both ppo and grpo training can run well after testing.

**Comment by hijkzzz** [comment]
The experience_maker and replay_buffer can also be merged?
Reducing duplicate code is beneficial for codebase maintenance.
And could you help to merge `n_responses` and `n_samples_per_prompt` ?

**Comment by hijkzzz** [comment]
n_samples_per_prompt has been fixed in https://github.com/OpenRLHF/OpenRLHF/commit/58c40ecc62c30f55df50cd3f2049c26efc0ded7f

**Comment by zhenghaoxu-gatech** [comment]
Hi, I want to confirm the implementation of `compute_approx_kl` in `openrlhf.models.utils`. The GRPO paper claims they use the unbiased estimator `pi_ref/pi - log(pi_ref/pi)-1` in http://joschu.net/blog/kl-approx.html. I see the blog is listed in codebase `openrlhf.models.utils` but the implementation seems to be the plain log-ratios. Shall we change it to be consistent with the GRPO paper?

**Comment by LSX-Sneakerprogrammer** [comment]
> Hi, I want to confirm the implementation of `compute_approx_kl` in `openrlhf.models.utils`. The GRPO paper claims they use the unbiased estimator `pi_ref/pi - log(pi_ref/pi)-1` in http://joschu.net/blog/kl-approx.html. I see the blog is listed in codebase `openrlhf.models.utils` but the implementation seems to be the plain log-ratios. Shall we change it to be consistent with the GRPO paper?

Hi, I am currently refactoring the code to make it cleaner. I notice the implementation you mentioned and will add it in the newest code, thanks for your suggestion

**Comment by LSX-Sneakerprogrammer** [comment]
Hi, I update the code to reduce its redundancy.  

The main ideas of this updates are as follows:
* update ```train_grpo_ray.py``` supporting ray and vllm for grpo.
* update ```Experience``` and ```BufferItem```, adding ```prompts``` for finding responses with same prompt, adding ```base_log_probs``` for calculating kl penalty for grpo later. Making ```values```, ```returns```, ```advantages```, ```base_log_probs```, ```prompts``` default to None, suitable for both GRPO and PPO training.
* split ```make_experience``` function of both ```NaiveExperienceMaker``` and ```RemoteExperienceMaker``` into ```make_experience``` and ```collect_base_experience```, which enables class ```GRPOExperienceMaker``` and ````GRPORemoteExperienceMaker``` to inherit the vast majority of the code collected from ```NaiveExperienceMaker``` and ```RemoteExperienceMaker```, in order to reduce code redundancy.
* update ```make_experience_batch``` and ```remove_padding_in_sequences```, which supports for both grpo and ppo 
* add ```GRPOReplayBuffer``` inherits from ```NaiveReplayBuffer```, which uses a dictionary ```prompts_dic``` to store responses with the same prompt, and split it into ```items``` by ```dic2items``` after calculating advatanges. Also supportting calculate advatanges between same prompts by ```calculate_advantages```
* add ```PPOActorModelRayActor``` and ```GRPOActorModelRayActor``` inherit from ```ActorModelRayActor``` in order to reduce code redundancy.
* add ```grpo_trainer.py```, ```grpo_actor.py```, ```train_grpo.py```, ```train_grpo_ray.py``` support for grpo training.
* update ```compute_approx_kl``` support for compute approximate kl, which mentioned in Schulman blog: http://joschu.net/blog/kl-approx.html and DeepSeekMath: https://arxiv.org/pdf/2402.03300

**Comment by hijkzzz** [comment]
Please rebase the MR to the main branch
Also cc @zhuzilin

**Comment by zhuzilin** [comment]
@LSX-Sneakerprogrammer 
Hi, after some offline discussion, it seems that PPO and GRPO can share lots of code and we have been working on preparing for such merge recently, by #463 and #466, where the first pr fixed the `n_samples_per_prompt`, and the second is trying to make "normalizing rewards" before calculating returns and advantages possible. They could provide a more general way to implement RL algorithms like PPO, GRPO, REINFORCE and PF-PPO.
Could you take a look at those PRs to see if that will make implementing GRPO easier? Sorry that this may have some large impact on the current implementation within this PR...

**Comment by LSX-Sneakerprogrammer** [comment]
> @LSX-Sneakerprogrammer Hi, after some offline discussion, it seems that PPO and GRPO can share lots of code and we have been working on preparing for such merge recently, by #463 and #466, where the first pr fixed the `n_samples_per_prompt`, and the second is trying to make "normalizing rewards" before calculating returns and advantages possible. They could provide a more general way to implement RL algorithms like PPO, GRPO, REINFORCE and PF-PPO. Could you take a look at those PRs to see if that will make implementing GRPO easier? Sorry that this may have some large impact on the current implementation within this PR...

Hi, thanks for your reply! The implementation in #463 PR has already considered, I modified all the codes based on #463. For #466, I see it is not difficult to change the implementaion based on this PR. Do I need to update the implemention based on #466 now or after #466 is merged into main branch.

**Comment by zhuzilin** [comment]
We'll try to merge #466 this week, so both will be fine.

BTW, I seems to me that we no longer need a separate `GRPOTrainer` or `ActorGRPOTrainer`, because we can reuse the code in PPO.
And also, could you try to rebase the code with main branch (or maybe merge the main)... because the current code diff is showing lots of changes from recent main commits... (I'm not sure if that's problem of the github code viewer...

**Comment by LSX-Sneakerprogrammer** [comment]
> We'll try to merge #466 this week, so both will be fine.
> 
> BTW, I seems to me that we no longer need a separate `GRPOTrainer` or `ActorGRPOTrainer` or `cli/train_grpo_ray.py`, because we can reuse the code in PPO. And also, could you try to rebase the code with main branch (or maybe merge the main)... because the current code diff is showing lots of changes from recent main commits... (I'm not sure if that's problem of the github code viewer...

Hi, I see the implementation in #466, it seems most of the GRPO codes can be reused from PPO, but some main changes cannot. For example, computing GRPO advatanges needs to normalize the rewards with the same prompts. Combined with the implementation in #466 , we can use an inherited class `GRPOTrainer` to rewrite the function ```get_advantages_and_returns```, which uses minimize codes. So I consider I can update the codes based on #466 PR for simplicity.

Meanwhile, I could merge the GRPO implementations in PPO, with adding `--use_grpo` as the suggestions from @hijkzzz 

BTW, for making the code diff minimum with main branch, maybe it's suitable to make a new PR based on #466?

**Comment by zhuzilin** [comment]
>  computing GRPO advatanges needs to normalize the rewards with the same prompts

The `process_experiences` method in #466 is added for doing such normalization.

**Comment by LSX-Sneakerprogrammer** [comment]
> > computing GRPO advatanges needs to normalize the rewards with the same prompts
> 
> The `process_experiences` method in #466 is added for doing such normalization.

Hi, I see the `process_experiences` method, I consider making changes based on #466 is more suitable. I will try to update code later.
